File: content/problems/0077-combinations/solutions/backtracking-pruning/Solution.java
class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> answer = new ArrayList<>(); backtrack(1, n, k, new ArrayList<>(), answer); return answer;
    }
    private void backtrack(int start, int n, int k, List<Integer> path, List<List<Integer>> answer) {
        if (path.size() == k) { answer.add(new ArrayList<>(path)); return; }
        int limit = n - (k - path.size()) + 1;
        for (int value = start; value <= limit; value++) { path.add(value); backtrack(value + 1, n, k, path, answer); path.remove(path.size() - 1); }
    }
}

File: content/problems/0077-combinations/solutions/backtracking-pruning/solution.cpp
class Solution {
    void backtrack(int start, int n, int k, vector<int>& path, vector<vector<int>>& answer) {
        if ((int)path.size() == k) { answer.push_back(path); return; }
        int limit = n - (k - path.size()) + 1;
        for (int value = start; value <= limit; ++value) { path.push_back(value); backtrack(value + 1, n, k, path, answer); path.pop_back(); }
    }
public:
    vector<vector<int>> combine(int n, int k) { vector<vector<int>> answer; vector<int> path; backtrack(1, n, k, path, answer); return answer; }
};
